man/dirfile2ascii.1
.\" dirfile2ascii.1.  The dirfile2ascii man page.
.\"
.\" Copyright (C) 2010, 2012 D. V. Wiebe
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"
.\" This file is part of the GetData project.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.2 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, with no Front-Cover Texts, and with no Back-Cover
.\" Texts.  A copy of the license is included in the `COPYING.DOC' file
.\" as part of this distribution.
.\"
.TH dirfile2ascii 1 "21 February 2012" "Version 0.8.0" "GETDATA"
.SH NAME
dirfile2ascii \(em output dirfile database vectors as ASCII text
.SH SYNOPSIS
.nh
.ad l
.TP
\fBdirfile2ascii\fR [ \fIOPTION\fR ]... \fIDIRFILE\fR
.RB "[ [ " \-a " | " \-A " | " \-e " | " \-E " | " \-F " | " \-g " | " \-G " |"
.BR \-o " | " \-i " | " \-u " | " \-x " | " \-X " ]"
.IR FIELD " ]..."
.hy
.ad n
.SH DESCRIPTION
Fetches data from a dirfile(5) database specified by
.I DIRFILE
and writes it as ASCII to standard output.  Any number of vector
.IR FIELD s
may be specified.  Each specified field is printed in a separate column.

Mandatory arguments to long options are mandatory for short options too.
.TP
.BR \-d ", " \-\-delimeter =\fIdelim\fR
separate columns by
.IR delim .
(Default: a single space.)
.TP
.BR \-f ", " \-\-first\-frame =\fIfirst_frame\fB\-\fIlast_frame\fR
read from frame
.I first_frame
to frame
.I last_frame
(inclusive).
.TP
.BR \-f ", " \-\-first\-frame =\fIfirst_frame\fB:\fInframes\fR
equivalent to \fB\-\-first\-frame\fR=\fIfirst_frame\fR
\fB\-\-num-frames\fR=\fInframes\fR.
.TP
.BR \-f ", " \-\-first\-frame =\fIfirst_frame\fR
If \fIfirst_frame\fR\~>=\~0, start reading at frame
.IR first_frame .
If 
.I first_frame
is -1 and \fB\-\-num-frames\fR=\fInframes\fR is specified, read the last
.I nframes
frames.  If
.B \-\-first-frame
is not specified, reading starts at frame zero.
.TP
.BR \-n ", " \-\-num\-frames =\fInframes\fR
read at most
.I nframes
frames.  If not specified, or if \fBnframes\fR\~=\~0, all frames to the
end-of-field are read.
.TP
.BR \-p ", " \-\-precision =\fIformat\fR
use
.I format
to format output.
.I format
may contain any of the flag characters, a field width, and/or a precision as
specified in
.BR printf (3).
It may \fInot\fR contain a length modifier.
.TP
.BR \-q ", " \-\-quiet
don't write diagnostic messages on standard error.  (This is the default
behaviour).
.TP
.BR \-s ", " \-\-skip =\fIframe_skip\fR
if \fIframe_skip\fR\~>\~0, output only one sample for every
.I frame_skip
frames.
.TP
.BR \-v ", " \-\-verbose
write diagnostic messages on standard error.
.TP
.BR \-z ", " \-\-fill =\fISTRING\fR
Fill columns which go past the end of their corresponding field with the string
.IR STRING .
The default behaviour is to fill columns with floating-point conversions with
.B NaN
and columns with integer conversion with
.BR 0 ,
which mirrors what occurs when an attempt is made to print data from before the
start of a field.  (Note: the default behaviour cannot be reproduced with this
option, since
.I STRING
is applied to all columns, regardless of conversion type.)
.P
In addition to the above, each
.I FIELD
argument may be preceded by a short option, one of:
.BR \-a ", " \-A ", " \-e ", " \-E ", " \-F ", " \-g ", " \-G ", " \-i ,
.BR \-o ", " \-u ", " \-x ", " \-X ,
indicating the conversion to be used.  See
.BR printf (3)
for the meaning of these conversion specifiers.  The output flags, width, and
precision may be specified by using
.BR \-\-precision .
If no conversion specifier is given,
.B %f
is used.

For conversion specifiers
.BR %a ", " %A ", " %e ", " %E ", " %f ", " %F ", " %g ", " %G ,
data is read from the dirfile as double precision floats.  For conversion
specifier \fB%i\fR, data is read as 64-bit signed integers.  For conversion
specifiers
.BR %o ", " %u ", " %x ", " %X ,
data is read as 64-bit unsigned integers.

.SH LIMITATIONS
No native support for printing complex data is provided.  This may be worked
around by using dirfile(5) representation suffixes.  For example, the command
.IP
.B dirfile2ascii
.I DIRFILE
.IB FIELD .r
.IB FIELD .i
.P
will print the real and imaginary parts of the complex valued field
.I FIELD
in the first and second columns, respectively.

.SH AUTHOR

dirfile2ascii was written by Matthew Truch and D. V. Wiebe.

.SH REPORTING BUGS

Please send reports of bugs to getdata\-devel@lists.sourceforge.net

The GetData home page: <http://getdata.sourceforge.net/>

.SH COPYRIGHT

Copyright \(co 2010 Matthew Truch.

dirfile2ascii is licenced under the GNU LPGL version 2.1 or later.

This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
.BR defile (1),
.BR dirfile (5),
.BR printf (3)


NEWS
New in version 0.8.0a:

  * Some stuff.

|==============================================================================|

New in version 0.7.3:

  Library Changes

  * BUG FIX: Passing a zero dividend to alter_crecip() and alter_recip() now
    results in the dividend not changing, per documentation.
  * BUG FIX: GetData wasn't properly computing the module directory, causing
    encodings using external modules to fail with GD_E_UNSUPPORTED.
  * BUG FIX: Metafield records are now properly stored in the format file for
    Standards Version 8.
  * BUG FIX: gd_raw_filename() wasn't clearing the Dirfile error before
    operation, resulting in it failing erroneously in certain situations.

  Utilities Changes:

  * BUG FIX: A potential segmentation fault has been patched in dirfile2ascii.

|==============================================================================|

New in version 0.7.2:

  Library Changes

  * If built with modules, the plugin shared objects will now be installed in
    a separate directory (by default "${libdir}/getdata", but see the
    --with-module-dir option in ./configure).
  * BUG FIX: When using modules, a missing plugin no longer results in a lock up
    the second time an attempt is made to find it.
  * BUG FIX: GetData now properly processes /INCLUDE directives with absolute
    paths, as the documentation insists it should.
  * BUG FIX: gd_add()ing the first raw field in a dirfile to a subfragment no
    longer results in memory corruption when calling gd_close().

  API Changes

  * If called with GD_CREAT but no encoding specified, gd_include() now will
    duplicate the encoding of the parent fragment (if any).

  Miscellaneous

  * The python module install dir can now be changed by passing
    --with-python-module-dir to configure.

|==============================================================================|

New in version 0.7.1:

  Library Changes

  * BUG FIX: When the parser fails in a gd_include() call, the internal list of
    conformant Standards Versions for the loaded dirfile is no longer corrupted.
  * BUG FIX Performing two consecutive reads of the same field in an ASCII
    encoded datasources no longer confuses GetData.
  * BUG FIX: Numerous memory leaks have been patched (mostly pin-holes, although
    there are a few, rare but egregious ones).

  API Changes:

  * The value(s) of a CONST or CARRAY field added with gd_[m]add() is now
    guaranteed to be zero; previously, it was unspecified.
  * BUG FIX: Specifying both GD_VERBOSE and GD_IGNORE_DUPS no longer results in
    spurious "Field code already defined" messages when duplicate fields are
    encountered.
  * BUG FIX: Calling gd_delete on a metafield no longer results in a segfault.

  Bindings Changes:

  * F77 BUG FIX: A memory leak in GDASCA has been fixed.
  * Python BUG FIX: The first element of a python list (instead of a NumPy
    array) is no longer dropped when passed to pygetdata on 64-bit systems.

  Utilities Changes:

  * dirfile2ascii now pads reads past the end-of-field.  The value used to
    pad may be specified with '-z', and defaults to 0/NaN.  Previously, the
    value of uninitialized memory was output.
  * BUG FIX: dirfile2ascii now does what the user intends when presented with
    cart-before-horse arguments, ie. "dirfile2ascii -e field DIRFILE".
    Previously this resulted in confusion.
  * BUG FIX: dirfile2ascii now accepts field conversion '-u', which was
    documented but overlooked in the argument parser.

|==============================================================================|

New in version 0.7.0:

  Dirfile Changes

  * Dirfile Standards Version 8 has been released.  It adds three new field
    types: DIVIDE, which computes X/Y for vectors X and Y, RECIP, which computes
    A/X for vector X and scalar A, and CARRAY, which is an array of CONST
    scalars.  A few other changes are listed below.
  * Like CONST fields, CARRAY elements can be used as parametersin the
    definition of other fields.  Syntax is 'field<n>', where 'field' is the
    field name and 'n' the element index in the CARRAY (and the angle brackets
    are literal angle brackets).  If the '<n>' is omitted, the first element
    (numbered 0) is assumed.
  * Slashes on reserved words are now mandatory, allowing fields with reserved
    words as names.
  * The long-deprecated single character data type codes have been removed from
    Standards Version 8.  (They are still accepted by GetData, which handles
    all previous Standards Versions as well.)
  * The /ENDIAN directive now takes an optional second parameter.  The only
    allowed value for this second token is "arm", which indicates that double
    precision floating point data (including double precision complex data) are
    stored in the ARM middle-endian format.  Without this, floating point data
    are assumed to have the same byte sex as integer data.

  Library Changes

  * GetData, including its bindings, is now supported under MacOS X, Cygwin, and
    Win32 using MinGW.  In the case of MinGW, this includes linking against the
    Microsoft C Runtime, which isn't POSIX compliant.  Users should excercise
    caution when writing to dirfiles in this case.
  * GetData can now be used on ARM processors.  As a side-effect, GetData can
    now read and write the middle-ended floating point format used by older ARM
    processors (on any processor, via endianness conversion).
  * Some changes have been made to the parser.  It now operates in one of two
    modes: pedantic mode, which adheres strictly to one particular Dirfile
    Standards Version, and permissive mode (the default) which is more
    lenient, as the old parser was.  Encountering a /VERSION directive will
    switch the parser from permissive to pedantic mode unless told not to (see
    GD_PERMISSIVE below), and will change the particular Standards Version to
    which pedantic mode is adhering.

util/dirfile2ascii.c
/* Copyright (C) 2010, 2011 Matthew Truch
 * Copyright (C) 2010 D. V. Wiebe
 *
 ***************************************************************************
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#define _XOPEN_SOURCE 1000
#define __STDC_FORMAT_MACROS
#define __STDC_LIMIT_MACROS
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <stdint.h>
#include <inttypes.h>
#include <getopt.h>
#include <math.h>
#include "nan.h"
#include "getdata.h"

#ifndef HAVE_OFF64_T
#define off64_t off_t
#endif

#define BUF_LEN GD_MAX_LINE_LENGTH
#define F_LEN 16
#define VALID_PRECISION_CHARS "#- +`I.0123456789"

struct field
{
  char *name;
  int type;
  gd_spf_t spf;
  union {
    double *dbl;
    int64_t *i64;
    uint64_t *u64;
  } u;
  char format[F_LEN];
  size_t n_read;
};

void version(void)
{
  printf("dirfile2ascii (%s)\n"
      "Copyright (C) 2010 Matthew Truch\n\n"
      "Please send reports of bugs and other communication to:\n  %s\n\n"
      "This program comes with NO WARRANTY, not even for MERCHANTABILITY "
      "or FITNESS\n"
      "FOR A PARTICULAR PURPOSE. You may redistribute it under the terms of "
      "the GNU\n"
      "Lesser General Public License, either version 2.1 of the License, or "
      "(at your\n"
      "option) any later version.\n\n"
      "You should have received a copy of the GNU Lesser General Public "
      "License along\n"
      "with this program; if not, write to the Free Software Foundation,"
      "Inc.,\n"
      "51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA\n",
      PACKAGE_STRING, PACKAGE_BUGREPORT);
  exit(-1);
}

void usage(char *exe)
{
  printf("Usage: %s [OPTION]... DIRFILE\n"
      "                    [[-a|-A|-e|-E|-F|-g|-G|-o|-i|-u|-x|-X] FIELD]...\n"
      "Dump specified fields from the dirfile DIRFILE as ASCII on stdout.\n\n"

      "Mandatory arguments to long options are mandatory for short options "
      "too.\n"
#if 0
      "  -b, --boxcar    boxcar average over skipped samples before output.  "
      "Ignored\n"
      "                    if --skip is not specified.\n"
#endif
      "  -d, --delimiter=D      separate columns by D (default: a single "
      "space)\n"
      "  -f, --first-frame=F-M  read from frame F to frame M (inclusive).\n"
      "  -f, --first-frame=F:N  equivalent to --first-frame=F --num-frames=N\n"
      "  -f, --first-frame=F    if F >= 0, start reading at frame F.  If F is "
      "-1 and\n"
      "                           --num-frames=N is specified, read the last N "
      "frames.\n"
      "                           If not specified, F = 0 is assumed.\n"
      "  -n, --num-frames=N     read at most N frames.  If not specified, or "
      "if N = 0,\n"
      "                           all frames to the end of the dirfile are "
      "read.\n"
      "  -p, --precision=FMT    use FMT to format output.  FMT may contain any "
      "of the\n"
      "                           flag characters, a field width, and/or a "
      "precision as\n"
      "                           specified in printf(3).  It may NOT contain "
      "a length\n"
      "                           modifier.\n"
      "  -q, --quiet            don't write verbose message to standard error "
      "(the\n"
      "                           default).\n"
      "  -s, --skip=S           if S > 0, output only one sample for every S "
      "frames.\n"
      "  -v, --verbose          write verbosely to standard error.\n"
      "  -z, --fill=STRING      print STRING to fill up columns which run past "
      "the end\n"
      "                           of their corresponding field.  The default "
      "is to\n"
      "                           fill floating point columns with \"NaN\" and "
      "integer\n"
      "                           columns with \"0\" (without quotation marks)."
      "\n\n"

      "Any number of fields may be specified.  Each specified field is printed "
      "in a\n"
      "separate column.  A field name may be preceded by a short option, one "
      "of -a, -A,\n"
      "-e, -E, -F, -g, -G, -o, -i, -u, -x, -X, indicating the conversion to be "
      "used.\n"
      "See printf(3) for the meaning of these conversion specifiers.  The "
      "output flags,\n"
      "width, and precision may be specified by --precision.  If no "
      "conversion\n"
      "specifier is given, %%f is used.\n\n"
      "For conversion specifiers %%a, %%A, %%e, %%E, %%f, %%F, %%g, %%G, data "
      "is read from the\n"
      "dirfile as double precision floats.  For conversion specifier %%i, the "
      "data is\n"
      "read as 64-bit signed integers.  For conversion specifiers %%o, "
      "%%u, %%x, %%X, the\n"
      "data is read as 64-bit unsigned integers.\n", exe);
  exit(-1);
}

int main (int argc, char **argv)
{
  char char_buffer[BUF_LEN];
  char *tmp;
  char *delimiter = " ";
  char *precision = "";
  char *zero = NULL;
  char *dirfile_name = NULL;
  DIRFILE *dirfile;
  size_t nf = 0;
  off64_t ff = 0;
  size_t n_want;
  gd_spf_t j;
  gd_spf_t max_spf = 0;
  gd_spf_t min_spf = UINT16_MAX;
  int c, i;
  int numfields = 0;
  struct field fields[BUF_LEN];
  int verbose = 0;
  int interpolate = 0;
  size_t skip = 0;
  int skipping = 0;
  size_t k, z;
#if 0
  int average = 0;
#endif
  int optind = 0;
  const double NaN = NAN;
  const struct option longopts[] = {
#if 0
    { "boxcar",      0, NULL, 'b' },
#endif
    { "delimieter",  1, NULL, 'd' },
    { "fill",        1, NULL, 'z' },
    { "first-frame", 1, NULL, 'f' },
    { "help",        0, NULL, '?' },
    { "num-frames",  1, NULL, 'n' },
    { "precision",   1, NULL, 'p' },
    { "quiet",       0, NULL, 'q' },
    { "skip",        1, NULL, 's' },
    { "verbose",     0, NULL, 'v' },
    { "version",     0, NULL,  2  },
    { 0,             0, 0,     0  }
  };

  /* handy things to know about conversion specifiers */
#define READ_AS_DOUBLE 0x000
#define READ_AS_INT    0x100
#define READ_AS_UINT   0x200
  struct {
    int t;
    const char* f;
  } type_data[0x7f];
  type_data[(int)'a'].t = READ_AS_DOUBLE;
  type_data[(int)'a'].f = "a";
  type_data[(int)'A'].t = READ_AS_DOUBLE;
  type_data[(int)'A'].f = "A";
  type_data[(int)'e'].t = READ_AS_DOUBLE;
  type_data[(int)'e'].f = "e";
  type_data[(int)'E'].t = READ_AS_DOUBLE;
  type_data[(int)'E'].f = "E";
  type_data[(int)'f'].t = READ_AS_DOUBLE;
  type_data[(int)'f'].f = "f";
  type_data[(int)'F'].t = READ_AS_DOUBLE;
  type_data[(int)'F'].f = "F";
  type_data[(int)'g'].t = READ_AS_DOUBLE;
  type_data[(int)'g'].f = "g";
  type_data[(int)'G'].t = READ_AS_DOUBLE;
  type_data[(int)'G'].f = "G";
  type_data[(int)'i'].t = READ_AS_INT;
  type_data[(int)'i'].f = PRIi64;
  type_data[(int)'o'].t = READ_AS_UINT;
  type_data[(int)'o'].f = PRIo64;
  type_data[(int)'u'].t = READ_AS_UINT;
  type_data[(int)'u'].f = PRIu64;
  type_data[(int)'x'].t = READ_AS_UINT;
  type_data[(int)'x'].f = PRIx64;
  type_data[(int)'X'].t = READ_AS_UINT;
  type_data[(int)'X'].f = PRIX64;

  while ((c = getopt_long(argc, argv,
          "-f:n:d:x:X:g:G:e:E:a:A:F:i:o:p:s:u:z:bvqh?", longopts, &optind)) != -1)
  {
    switch (c) {
      case 1: /* The case of no option */
        if (dirfile_name == NULL) {
          dirfile_name = optarg;
        } else { /* Standard output field */
          if (numfields > BUF_LEN) {
            fprintf(stderr, "Error: Too many fields!\n");
            exit(-2);
          }
          fields[numfields].type = 'f';
          fields[numfields].name = optarg;
          numfields++;
        }
        break;
      case 2:
        version();
        break;
      case 'x': /* output field types */
      case 'X':
      case 'i':
      case 'o':
      case 'g':
      case 'G':
      case 'e':
      case 'E':
      case 'u':
      case 'F':
      case 'A':
      case 'a':
        if (numfields > BUF_LEN) {
          fprintf(stderr, "Error: Too many fields!\n");
          exit(-2);
        }
        fields[numfields].type = c;
        fields[numfields].name = optarg;
        numfields++;
        break;
      case 'f': /* firstframe */
        ff = (off64_t)strtoll(optarg, &tmp, 0);
        if (tmp[0] == ':' && tmp[1] != '\0')
          nf = (size_t)strtoll(&(tmp[1]), NULL, 0);
        else if (tmp[0] == '-' && tmp[1] != '\0')
          nf = (size_t)(strtoll(&(tmp[1]), NULL, 0) - ff);
        break;
      case 'n': /* numframes */
        nf = (size_t)strtoll(optarg, NULL, 0);
        break;
      case 'p':
        if (strlen(optarg) < (F_LEN - 6))
          precision = optarg;
        else
          fprintf(stderr, "Ignoring exceedingly long -p argument (%s).\n",
              optarg);
        break;
      case 'v':
        verbose = 1;
        break;
      case 'q':
        verbose = 0;
        break;
      case 'd':
        delimiter = optarg;
        break;
      case 's':
        skip = strtoll(optarg, NULL, 0);
        break;
      case 'z':
        zero = optarg;
        break;
      case 'b':
#if 0
        average = 1;
#endif
        fprintf(stderr, "Warning: Ignoring -b found on command line.\n");
        break;
      case 'h':
      case '?':
      default:
        usage(argv[0]);
        break;
    }
  }

  if (dirfile_name == NULL) {
    fprintf(stderr, "Error: No dirfile specified.\n");
    exit(-4);
  }

  if (numfields < 1) {
    fprintf(stderr, "Error: At least one field must be specified.\n");
    exit(-3);
  }

  if ((ff == -1) && (nf == 0)) {
    fprintf(stderr, "Warning: Ignoring negative first frame with no frame "
        "count given.\n");
    ff = 0;
  }

  for (z = 0; z < strlen(precision); z++) {
    if (strchr(VALID_PRECISION_CHARS, precision[z]) == NULL) {
      fprintf(stderr,
          "Error: Invalid character (%c) found in precision string (%s).\n", 
          precision[z], precision);
      exit(-5);
    }
  }

  if (skip)
    skipping = 1;
  if (skip < 1)
    skip = 1;

  dirfile = gd_open(dirfile_name, GD_RDONLY);
  if (gd_error(dirfile)) {
    fprintf(stderr, "GetData error: %s\n", gd_error_string(dirfile, char_buffer,
          BUF_LEN));
    gd_close(dirfile);
    exit(1);
  }

  if (nf == 0) {
    nf = gd_nframes(dirfile) - ff;
    if (gd_error(dirfile)) {
      fprintf(stderr, "GetData error: %s\n", gd_error_string(dirfile,
            char_buffer, BUF_LEN));
      gd_close(dirfile);
      exit(3);
    }
  }

  if (ff == -1) {
    ff = gd_nframes(dirfile) - nf;
    if (gd_error(dirfile)) {
      fprintf(stderr, "GetData error: %s\n", gd_error_string(dirfile,
            char_buffer, BUF_LEN));
      gd_close(dirfile);
      exit(3);
    }
  }

  /* Get spfs and sanity checks for all fields */
  for (i = 0; i < numfields; i++) {
    fields[i].spf = gd_spf(dirfile, fields[i].name);
    if (gd_error(dirfile)) {
      fprintf(stderr, "GetData error: %s\n", gd_error_string(dirfile,
            char_buffer, BUF_LEN));
      gd_close(dirfile);
      exit(3);
    }

    if (max_spf && fields[i].spf != max_spf)
      interpolate = 1;

    if (fields[i].spf > max_spf)
      max_spf = fields[i].spf;

    if (fields[i].spf < min_spf)
      min_spf = fields[i].spf;
  }

  if (interpolate && nf == 1 && min_spf == 1) {
    fprintf(stderr, "Error: Interpolation required, "
        "but at least one field has only 1 sample.\n");
    exit(-6);
  }

  if (verbose) {
    fprintf(stderr, "Reading %d field%s from %s\n", numfields,
        (numfields > 1) ? "s" : "", dirfile_name);
    fprintf(stderr, "First frame: %llu Number of frames: %zd\n",
        (long long)ff, nf);
  }

  for (i = 0; i < numfields; i++) { /* Read in all the fields */
    n_want = nf * fields[i].spf;

    if (type_data[fields[i].type].t == READ_AS_DOUBLE) {
      fields[i].u.dbl = (double *)malloc(sizeof(double) * n_want);
      if (fields[i].u.dbl == NULL) {
        perror("malloc");
        gd_close(dirfile);
        exit(4);
      }

      fields[i].n_read = gd_getdata(dirfile, fields[i].name, ff, 0, nf, 0,
          GD_FLOAT64, fields[i].u.dbl);
    } else if (type_data[fields[i].type].t == READ_AS_INT) {
      fields[i].u.i64 = (int64_t *)malloc(sizeof(int64_t) * n_want);
      if (fields[i].u.i64 == NULL) {
        perror("malloc");
        gd_close(dirfile);
        exit(4);
      }
      fields[i].n_read = gd_getdata(dirfile, fields[i].name, ff, 0, nf, 0,
          GD_INT64, fields[i].u.i64);
    } else {
      fields[i].u.u64 = (uint64_t *)malloc(sizeof(uint64_t) * n_want);
      if (fields[i].u.u64 == NULL) {
        perror("malloc");
        gd_close(dirfile);
        exit(4);
      }
      fields[i].n_read = gd_getdata(dirfile, fields[i].name, ff, 0, nf, 0,
          GD_UINT64, fields[i].u.u64);
    }

    if (gd_error(dirfile)) {
      fprintf(stderr, "GetData error: %s\n", gd_error_string(dirfile,
            char_buffer, BUF_LEN));
      gd_close(dirfile);
      exit(5);
    } else if (!zero && n_want > fields[i].n_read) {
      if (verbose)
        fprintf(stderr, "Short read on field %i, padding %zu frames\n",
            i, n_want - fields[i].n_read);
      if (type_data[fields[i].type].t == READ_AS_DOUBLE) {
        for (k = fields[i].n_read; k < n_want; ++k)
          fields[i].u.dbl[k] = NaN;
      } else if (type_data[fields[i].type].t == READ_AS_INT) {
        memset(fields[i].u.i64 + fields[i].n_read, 0,
            (n_want - fields[i].n_read) * sizeof(int64_t));
      } else {
        memset(fields[i].u.u64 + fields[i].n_read, 0,
            (n_want - fields[i].n_read) * sizeof(uint64_t));
      }
    }
  }

  /* Generate format string with precision */
  for (i = 0; i < numfields; i++)
    snprintf(fields[i].format, F_LEN, "%%%s%s", precision,
        type_data[fields[i].type].f);

  for (k = 0; k < nf; k += skip) {
    for (j = 0; j < (skipping ? 1 : max_spf); j++) {
      for (i = 0; i < numfields; i++) {
        if (fields[i].spf == max_spf || skipping) {
          if (zero && k * fields[i].spf + j >= fields[i].n_read)
            printf("%s", zero);
          else if (type_data[fields[i].type].t == READ_AS_DOUBLE)
            printf(fields[i].format, fields[i].u.dbl[k * fields[i].spf + j]);
          else if (type_data[fields[i].type].t == READ_AS_INT)
            printf(fields[i].format, fields[i].u.i64[k * fields[i].spf + j]);
          else
            printf(fields[i].format, fields[i].u.u64[k * fields[i].spf + j]);
        } else { /* need to interpolate: */
          /*      formula is  y = y0 + (x-x0) * (y1-y0)/(x1-x0) */
          /*                 val= y0 +  diff  *     slope       */
          /* Note that if max_spfs isn't a multiple of all spfs then some of  */
          /* the data will only be approximated (except at frame boundaries). */
          gd_spf_t prev_samp, next_samp, offset;
          double slope, diff, val;

          prev_samp = (gd_spf_t)floor((double)j * (double)fields[i].spf /
              (double)max_spf);
          next_samp = prev_samp + 1;
          diff = ((double)prev_samp + (double)(j % (max_spf / fields[i].spf)) /
              (double)(max_spf / fields[i].spf)) - (double)prev_samp;

          if (k == (nf - 1) && next_samp == fields[i].spf)
            /* we're on ultimate sample, so use slope from previous set */
            offset = 1;
          else
            offset = 0;

          if (zero && k * fields[i].spf + j >= fields[i].n_read) {
            printf("%s", zero);
          } else if (type_data[fields[i].type].t == READ_AS_DOUBLE) {
            slope = (fields[i].u.dbl[k * fields[i].spf + next_samp - offset] -
                fields[i].u.dbl[k * fields[i].spf + prev_samp - offset]) /
              ((double)next_samp - (double)prev_samp);
            val = fields[i].u.dbl[k * fields[i].spf + prev_samp] + diff * slope;
            printf(fields[i].format, val);
          } else if (type_data[fields[i].type].t == READ_AS_INT) {
            slope = ((double)fields[i].u.i64[k * fields[i].spf + next_samp
                - offset] - (double)fields[i].u.i64[k * fields[i].spf
                + prev_samp - offset]) / ((double)next_samp
                  - (double)prev_samp);
            val = (double)fields[i].u.i64[k * fields[i].spf + prev_samp] + diff
              * slope;
            printf(fields[i].format, (int64_t)val);
          } else {
            slope = ((double)fields[i].u.u64[k * fields[i].spf + next_samp
                - offset] - (double)fields[i].u.u64[k * fields[i].spf
                + prev_samp - offset]) / ((double)next_samp
                  - (double)prev_samp);
            val = (double)fields[i].u.u64[k * fields[i].spf + prev_samp] + diff
              * slope;
            printf(fields[i].format, (uint64_t)val);
          }
        }

        if (i < (numfields - 1)) {
          printf("%s", delimiter);
        } else {
          printf("\n");
        }
      }
    }
  }

  gd_close(dirfile);
  return 0;
}
